### PR TITLE
Connections crud

### LINKED ISSUE #7576: [Server] Add & expose Connections management capability
#### Current Behavior

No connections API

#### Desired Behavior

- Adding Connection persistence capability to Meshery server.
- Exposing connections capability to UI via API.

#### Implementation

API endpoint `/api/integrations/connections`

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/contributing/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by Azanul**
**Notes for Reviewers**

This PR fixes #7576 

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [X] Yes, I signed my commits.

### CODE REVIEW

File: server/handlers/connections_handlers.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	h.log.Info("connection saved successfully")
+	w.WriteHeader(http.StatusCreated)
+}
+
+func (h *Handler) GetConnections(w http.ResponseWriter, req *http.Request, _ *models.Preference, user *models.User, provider models.Provider) {
+	q := req.URL.Query()
+	pathSegments := strings.Split(req.URL.Path, "/")
+	connectionKind := pathSegments[len(pathSegments)-1]
```
Comment (MUzairS15) [inline server/handlers/connections_handlers.go:45]: Instead of string interpolation, why not use `mux.Vars(r)["connectionKind"]`.

**Review by MUzairS15 — CHANGES_REQUESTED** [review]

**Comment by Chadha93** [comment]
@Azanul Let's discuss this on the Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

**Comment by leecalcote** [comment]
I missed it. Was this demo'ed last week? Demo again tomorrow, if so? What's preventing its merge? @Azanul @harkiratsm

**Review by MUzairS15 — COMMENTED** [review]
Docs update are missing

**Review by leecalcote — CHANGES_REQUESTED** [review]
A couple lint checks are in need of attention.

**Comment by Azanul** [comment]
> A couple lint checks are in need of attention.

You mean the failing codecov checks?

**Comment by l5io** [comment]
🚀 Preview for commit 67eb99310218f9f133eb856216873cfb6b3ec851 at: https://64593eca6b85f21482dd95f2--meshery-docs-preview.netlify.app

**Review by MUzairS15 — CHANGES_REQUESTED** [review]
I don’t see docs  updated in preview build, did you ran make docs-build ??

**Comment by leecalcote** [comment]
> I don’t see docs updated in preview build, did you ran make docs-build ??

@Azanul still working on this?

**Comment by Azanul** [comment]
> @Azanul still working on this?

Need to update the implementation and the docs according to [the discussion](https://layer5io.slack.com/archives/G019GNQGPL4/p1683646920505949).

**Comment by leecalcote** [comment]
> > @Azanul still working on this?
> 
> Need to update the implementation and the docs according to [the discussion](https://layer5io.slack.com/archives/G019GNQGPL4/p1683646920505949).

Ok. Please send word when ready for re-review.

**Review by theBeginner86 — COMMENTED** [review]
@Azanul 
Will you please make the necessary changes based on what API structure we have?

@harkiratsm's https://github.com/meshery/meshery/pull/7634 is too dependent on this.

**Comment by l5io** [comment]
🚀 Preview for commit 34034141d0b99dbf36f009afd6eb3575fcfef5a7 at: https://6474d937ce74732dc73ac65f--meshery-docs-preview.netlify.app

**Comment by leecalcote** [comment]
@Azanul please take note of the codecov comments.

**Review by theBeginner86 — COMMENTED** [review]
Looks good.

Would you please see why docs are not generated? 
There are missing swagger comments on the handler functions.

**Comment by Azanul** [comment]
> Looks good.
> 
> Would you please see why docs are not generated? There are missing swagger comments on the handler functions.

Check it out

**Comment by l5io** [comment]
🚀 Preview for commit 613c355dfe7c831d892c813b2a1fb6fe5821c318 at: https://6476fe24350b57780a8900b2--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit 24a90148197d72dd8619aa3a451ab4736912adfa at: https://647707c69e75147c21c5b2f6--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit adca50b35c43e06f68c2f9f0c21535c68eb4c65c at: https://64770f672bbffa0a53a5725d--meshery-docs-preview.netlify.app

**Comment by leecalcote** [comment]
@MUzairS15, I wonder if you might have any comments to offer here.

**Comment by l5io** [comment]
🚀 Preview for commit 54220d32bc7adaa3701f8f5cf7e2a86111bb8951 at: https://64777f817735d80080d9d052--meshery-docs-preview.netlify.app

**Comment by l5io** [comment]
🚀 Preview for commit a69c9ed51d87a956fae8a95a4b4a054c1dede3d1 at: https://64778bfa8b4f1d00b89ea478--meshery-docs-preview.netlify.app

File: server/router/server.go [hunk 2]
```diff
@@ -305,6 +305,16 @@ func NewRouter(_ context.Context, h models.HandlerInterface, port int, g http.Ha
 	gMux.Handle("/api/integrations/credentials", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.DeleteUserCredential), models.ProviderAuth))).
 		Methods("DELETE")
 
+	// Handlers for User Connections
+	gMux.Handle("/api/integrations/connection", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.SaveConnection), models.ProviderAuth))).
+		Methods("POST")
+	gMux.Handle("/api/integrations/connections/{connectionKind}", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.GetConnections), models.ProviderAuth))).
+		Methods("GET")
+	gMux.Handle("/api/integrations/connection/{connectionKind}/edit", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.UpdateConnection), models.ProviderAuth))).
```
Comment (abdullah1308) [inline server/router/server.go:313]: Do we really need `/edit` ? We can have the same endpoint with the `PUT` verb. Similarly for `/delete`

File: server/handlers/connections_handlers.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	if err := json.NewEncoder(w).Encode(connectionsPage); err != nil {
+		h.log.Error(fmt.Errorf("error encoding user connections: %v", err))
+		http.Error(w, "unable to encode user connections", http.StatusInternalServerError)
+		return
+	}
+}
+
+// swagger:route PUT /api/integrations/connection/{connectionKind}/edit PutConnection idPUtConnection
```
Comment (abdullah1308) [inline server/handlers/connections_handlers.go:97]: rename to `idPutConnection`

File: server/router/server.go [hunk 4]
```diff
@@ -305,6 +305,16 @@ func NewRouter(_ context.Context, h models.HandlerInterface, port int, g http.Ha
...
+		Methods("PUT")
+	gMux.Handle("/api/integrations/connection/{connectionKinId}/delete", h.ProviderMiddleware(h.AuthMiddleware(h.SessionInjectorMiddleware(h.DeleteConnection), models.ProviderAuth))).
```
Comment (abdullah1308) [inline server/router/server.go:315]: rename to `connectionId`

File: server/handlers/connections_handlers.go (hunk 3 above)
Comment (abdullah1308) [inline server/handlers/connections_handlers.go]: rename to `IdPutConnection`

File: server/router/server.go (hunk 2 above)
Comment (Azanul) [reply server/router/server.go:313]: > Do we really need /edit ?

We don't.

I think I kept it like that to keep 1:1 correlation with Meshery Cloud

File: server/handlers/connections_handlers.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+// swagger:route DELETE /api/integrations/connection/{connectionId}/delete DeleteConnection idDeleteConnection
+// Handle DELETE request for deleting an existing connection by connection ID
+//
+// Deletes existing connection
+// responses:
+// 200: noContentWrapper
+func (h *Handler) DeleteConnection(w http.ResponseWriter, req *http.Request, _ *models.Preference, _ *models.User, provider models.Provider) {
+	q := req.URL.Query()
+
+	connectionID := uuid.FromStringOrNil(q.Get("connection_id"))
```
Comment (abdullah1308) [inline server/handlers/connections_handlers.go:139]: rename to `connectionId`

**Comment by l5io** [comment]
🚀 Preview for commit 3e9095de4fea8891d8e38231aa713ee9cc4e7282 at: https://647de2c01fa52e0da47941cf--meshery-docs-preview.netlify.app

File: server/models/remote_provider.go [hunk 6]
```diff
@@ -3463,6 +3463,131 @@ func (l *RemoteProvider) SaveConnection(req *http.Request, conn *Connection, tok
...
+		return nil, ErrInvalidCapability("PersistConnection", l.ProviderName)
+	}
+	ep, _ := l.Capabilities.GetEndpointForFeature(PersistConnection)
+	_creds, err := json.Marshal(connection)
+	if err != nil {
+		return nil, err
+	}
+	bf := bytes.NewBuffer(_creds)
+	remoteProviderURL, _ := url.Parse(fmt.Sprintf("%s%s/%s/edit", l.RemoteProviderURL, ep, connection.Kind))
+	cReq, _ := http.NewRequest(http.MethodPost, remoteProviderURL.String(), bf)
```
Comment (abdullah1308) [inline server/models/remote_provider.go:3519]: change to `http.MethodPut`

**Review by abdullah1308 — CHANGES_REQUESTED** [review]

**Comment by abdullah1308** [comment]
@Azanul

File: server/router/server.go (hunk 4 above)
Comment (abdullah1308) [reply server/router/server.go:315]: @Azanul I've just checked it. It is `connectionId`

**Comment by l5io** [comment]
🚀 Preview for commit e90f840ffeba36819db4ccbb69f7cd935fa88a70 at: https://647de68bed84ff0e32edc1b7--meshery-docs-preview.netlify.app

File: server/router/server.go (hunk 4 above)
Comment (Azanul) [reply server/router/server.go:315]: @abdullah1308 Now I see what you mean. I thought you were talking about the PUT request. I'll update it right away.

**Review by abdullah1308 — APPROVED** [review]

**Comment by abdullah1308** [comment]
@theBeginner86 review and merge

**Comment by l5io** [comment]
🚀 Preview for commit b545806dc2fc4a419245f139a5a03a80e6751ac1 at: https://647e0bb7629b8a00ae8a0458--meshery-docs-preview.netlify.app

**Comment by abdullah1308** [comment]
@MarioArriaga92 can you look into the failing UI integration test ?

**Comment by MarioArriaga92** [comment]
@abdullah1308 @leecalcote this is the output for the failing integration cypress test hadn't seen this before but will look into it on master branch locally and other PR CI runs to identify why & how frequent it happens:
CleanShot 2023-06-05 at 12 56 04

**Pull request merged**